k8s: Provide fallback for EndpointSlices detection if discovery API is not available

tgraf opened the pull request:
Discovery of API groups requires the API services of the apiserver to be healthy. Such API services can depend on the readiness of regular pods which require Cilium to function correctly. By treating failure to discover API groups as fatal, a critial loop can be entered in which Cilium cannot start because the API groups can't be discovered and the API groups will only become discoverable once Cilium is up.

Warn about the lack of discovery ability and fall back to probing the API directly.

errordeveloper reviewed (approved):
LGTM 👍 

nit: could you clarify if `Update` would get called again? The description comment for this function doesn't necessarily suggest what is the case... Perhaps copying the info from commit message into a comment would be quite helpful also, especially as this is a pretty special case.

aanm reviewed (changes requested):
Discussed with Thomas offline. This requires a flag that does the discovery on a conditional basis

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -130,22 +137,64 @@ func Force(version string) error {
-		return err
+		// It doesn't make sense to retry the retrieval of this
+		// information at a later point because the capabilities are
+		// primiarly used while the agent is starting up. Instead, fall
```
`s/primiarly/primarily`

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -130,22 +137,64 @@ func Force(version string) error {
+	// Discovery of API groups requires the API services of the apiserver
+	// to be healthy. Such API services can depend on the readiness of
+	// regular pods which require Cilium to function correctly. By treating
+	// failure to discover API groups as fatal, a critial loop can be
```
`s/critial/critical``

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -130,22 +137,64 @@ func Force(version string) error {
 // capabilities. This function must be called after connectivity to the
 // apiserver has been established.
+//
+// Discovery of capabilities only works if the disocovery API of the apiserver
```
`s/disocovery/discovery`

aanm commented on pkg/k8s/config/config.go:
```diff
@@ -0,0 +1,40 @@
+// K8sAPIDiscoveryEnabled returns true if API discovery of API groups and
+// resources is enabled
+func (d *DefaultConfiguration) K8sAPIDiscoveryEnabled() bool {
+	return defaults.K8sEnableAPIDiscovery
```
Is this right? Shouldn't this be the value specified by the user?

tgraf replied:
Yes, this represents the default configuration when no user options are available. This also clearly documents in which cases that happens instead of just relying on `pkg/option` which is uninitialized in several cases.

aanm reviewed (changes requested):
To be fair I only noticed the typos because I opened the PR on my IDE.

Weil0ng commented on pkg/k8s/version/version.go:
```diff
@@ -130,22 +137,64 @@ func Force(version string) error {
+		}
+	}
+
+	// Unknown error, we can't derive whethr to enable or disable
```
nit: s/whethr/whether

Weil0ng commented on pkg/k8s/version/version.go:
```diff
@@ -17,17 +17,24 @@
 
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "k8s")
+
 // ServerCapabilities is a list of server capabilities derived based on version
```
nit: might change the description here since it is no longer *just* based on version?

joestringer reviewed (approved):
CLI/docs changes LGTM.

tgraf committed:
k8s: Fallback to API probing if discovery API is not available

Discovery of API groups requires the API services of the apiserver to be
healthy. Such API services can depend on the readiness of regular pods
which require Cilium to function correctly. By treating failure to
discover API groups as fatal, a critial loop can be entered in which
Cilium cannot start because the API groups can't be discovered and th
API groups will only become discoverable once Cilium is up.

Warn about the lack of discovery ability and fall back to probing the
API directly.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Require use of discovery API to be enabled

The discovery request consists of several API calls to the apiserver. It
has already been observed that the k8s client can stsrt throttling
during startup phase.

Disable the use of the discovery API by default and rely on individual
API probing but allow users to benefit from it by opting into it.

Signed-off-by: Thomas Graf <thomas@cilium.io>
